### PR TITLE
#466: Kill running Spark Job

### LINKED ISSUE #466: Kill running Spark Job
Once spark job is submitted
User should be able to see kill job button on runs detail page and
should be able to kill running spark job

On the backed side we should use yarn application id and YARN REST API: 
http://hadoop.apache.org/docs/r2.5.2/hadoop-yarn/hadoop-yarn-site/ResourceManagerRest.html#Cluster_Application_State_API

### PR DESCRIPTION
**Pull request by jozefbakus**
Implements: #466 - Kill running Spark Job

User is only able to kill Running Spark job with application id from yarn

New application property:
sparkYarnSink.userUsedToKillJob

### CODE REVIEW

File: src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/controllers/JobInstanceController.scala [hunk 1]
```diff
@@ -33,5 +32,10 @@ class JobInstanceController @Inject()(jobInstanceService: JobInstanceService) {
     jobInstanceService.getJobInstances(dagInstanceId).toJava.toCompletableFuture
   }
 
+  @PostMapping(path = Array("/jobInstances"))
```
Comment (kevinwallimann) [inline src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/controllers/JobInstanceController.scala]: We need to clarify anyway how we want to deal with custom actions (search, activate, kill), but for the time being, it's pretty surprising that POST /jobInstances leads to killing 1 job. 1. It should not be plural, 2. POST /entityName usually maps to creating the object. I'd use `/jobInstance/kill` for the moment

File: src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/services/JobInstanceService.scala [hunk 2]
```diff
@@ -33,4 +39,19 @@ class JobInstanceServiceImpl(override val jobInstanceRepository: JobInstanceRepo
...
+    val url: String =
+      s"${SparkExecutorConfig.getHadoopResourceManagerUrlBase}/ws/v1/cluster/apps/$applicationId/state?user.name=${SparkExecutorConfig.getUserUsedToKillJob}"
+    val data = Json.obj(
+      "state" -> "KILLED"
+    )
+
+    Try {
+      WSClientProvider.getWSClient.url(url).put(data).map { response =>
+        response.status == 200 || response.status == 202
+      }.recoverWith {
```
Comment (kevinwallimann) [inline src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/services/JobInstanceService.scala:52]: Nice, so there's no need to update any status, because the status is updated anyway by the usual scheduler loop. That's really nice and decoupled

File: ui/src/app/constants/texts.constants.ts [hunk 3]
```diff
@@ -51,6 +51,12 @@ export const texts = {
   RUN_WORKFLOWS_SUCCESS_NOTIFICATION: 'Workflows have been submitted.',
   RUN_WORKFLOWS_FAILURE_NOTIFICATION: "Sorry, workflows couldn't be submitted. Please try again.",
 
+  KILL_JOB_CONFIRMATION_TITLE: 'Kill running job',
+  KILL_JOB_CONFIRMATION_CONTENT:
+    'Are you sure you want to kill running job? The operation cannot be reverted and operation may not take effect.',
```
Comment (kevinwallimann) [inline ui/src/app/constants/texts.constants.ts]: This is correct, but it sounds a bit confusing. I would only keep "The operation cannot be reverted" and omit "operation may not take effect"

File: ui/src/app/constants/texts.constants.ts [hunk 4]
```diff
@@ -51,6 +51,12 @@ export const texts = {
...
+  KILL_JOB_SUCCESS_NOTIFICATION: 'Kill job request passed, but it does not mean that job is killed. Please refresh page and verify status.',
```
Comment (kevinwallimann) [inline ui/src/app/constants/texts.constants.ts]: I'd rephrase "Request to kill job passed" Otherwise it is a bit confusing it could mean killing a job request (whatever a job request is) or requesting to kill the job. Again it is 100% correct, but a bit confusing. How about

"Request to kill the job passed. Please refresh the page after a few seconds and verify its status"

File: src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/services/JobInstanceService.scala (hunk 2 above)
Comment (jozefbakus) [reply src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/services/JobInstanceService.scala:52]: Exactly

File: ui/src/app/constants/texts.constants.ts (hunk 4 above)
Comment (jozefbakus) [reply ui/src/app/constants/texts.constants.ts]: Done

File: ui/src/app/constants/texts.constants.ts (hunk 3 above)
Comment (jozefbakus) [reply ui/src/app/constants/texts.constants.ts]: Ok

File: src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/controllers/JobInstanceController.scala (hunk 1 above)
Comment (jozefbakus) [reply src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/controllers/JobInstanceController.scala]: I will follow the current convention and we can start thinking about REST API refactoring and follow rest paradigm

**Review by kevinwallimann — APPROVED** [review]
